src/tvheadend/HTSPVFS.cpp

```diff
@@ -108,15 +116,58 @@ ssize_t HTSPVFS::Read ( unsigned char *buf, unsigned int len )
   if (read > 0)
     m_offset += read;
 
+  /* Tvheadend may briefly return 0 bytes when playing an in-progress recording at end-of-file
+     we'll retry 50 times with 10ms pauses (~500ms) before giving up */
+  if (read == 0 && inprogress)
+  {
+    for (int i = 1; i <= 50; i++)
+    {
+      read = SendFileRead(buf, len);
+      if (read > 0)
+      {
+        m_offset += read;
+        return read;
+      }
+      usleep(10000);
+    }
+    Logger::Log(LogLevel::LEVEL_ERROR, "vfs inprogress recording read failed after 50 attempts");
```

Comment:
Is this an error or a warning? I think it is more a warning.

Reply:
Could't this method be simplified to:

```c++
{
  /* Not opened */
  if (!m_fileId)
    return -1;

  /* Tvheadend may briefly return 0 bytes when playing an in-progress recording at end-of-file
     we'll retry 50 times with 10ms pauses (~500ms) before giving up */
  int tries = inprogress ? 1 : 50;

  for (int i = 1; i <= tries; i++)
  {
    ssize_t read = SendFileRead(buf, len);
    if (read > 0)
    {
      m_offset += read;
      return read;
    }
    else if (i < tries)
      usleep(10000);
  }
  Logger::Log(LogLevel::LEVEL_WARNING, "vfs inprogress recording read failed after %d attempts", tries);
  return -1;
}
```

(not tested!)

Reply:
That's certainly cleaner code... will adjust the PR

Had thought about doing it that way; think I was trying to make the intention of the loop clear. That said, I later added the comment which explains things.

Will adjust the log message too as this makes it (potentially) apply to normal recordings too.

Not too sure about returning -1 when (final) read could be 0 - this is a change in current behaviour.

Shouldn't we return whatever was returned by the previous SendFileRead like it does currently?

Reply:
Hows this look

```cpp
{
  /* Not opened */
  if (!m_fileId)
    return -1;

  /* Tvheadend may briefly return 0 bytes when playing an in-progress recording at end-of-file
     we'll retry 50 times with 10ms pauses (~500ms) before giving up */
  int tries = inprogress ? 1 : 50;
  ssize_t read;

  for (int i = 1; i <= tries; i++)
  {
    read = SendFileRead(buf, len);
    if (read > 0)
    {
      m_offset += read;
      return read;
    }
    else if (i < tries)
      usleep(10000);
  }
  Logger::Log(LogLevel::LEVEL_WARNING, "vfs read failed after %d attempts", tries);
  return read;
}
```

Reply:
Looking good, except the not initialzed 'read' variable. if somebody chges the code so that the loop might be not executed at all we will end with a random return value.

Reply:
Agreed will start it out at zero.

However, not currently able to compile as LEVEL_WARNING is not declared....

```
pvr.hts/src/tvheadend/HTSPVFS.cpp: In member function 'ssize_t tvheadend::HTSPVFS::Read(unsigned char*, unsigned int, bool)':
pvr.hts/src/tvheadend/HTSPVFS.cpp:127:15: error: 'LEVEL_WARNING' is not a member of 'tvheadend::utilities::LogLevel'
```

Looks like we only have these declared in Logger.h:
      LEVEL_ERROR,
      LEVEL_INFO,
      LEVEL_DEBUG,
      LEVEL_TRACE

Shall I go for LEVEL_DEBUG / INFO or extend "enum LogLevel" in src/tvheadend/utilities/Logger.h to include LEVEL_WARNING?

Reply:
I suggest to go far DEBUG.

Reply:
Done

Reply:
Just found this obvious mistake so will correct this:
`int tries = inprogress ? 1 : 50;`

Should be:
`int tries = inprogress ? 50 : 1;`

Will update the PR shortly.

Reply:
Now corrected